File: 09-backtracking/00078-Subsets/00078-subsets.go
// 78: Subsets
// https://leetcode.com/problems/subsets/

package main

import "fmt"


func backtrack(nums []int, start int, current []int, result *[][]int) {
	*result = append(*result, current)

	for i:=start; i<len(nums); i++ {
		current = append(current, nums[i])
		backtrack(nums, (i + 1), current, result);
		current = current[:len(current)-1]
	}
}


// SOLUTION
func subsets(nums []int) [][]int {
    var current []int
	var result [][]int
	backtrack(nums, 0, current, &result);
	return result;
}

func main() {
    // INPUT
    nums := []int{1,2,3}

    // OUTPUT
    result := subsets(nums);

	fmt.Println(result)
}

File: 09-backtracking/00078-Subsets/00078-subsets.java
// 78: Subsets
// https://leetcode.com/problems/subsets/


import java.util.List;
import java.util.ArrayList;

class Solution {
    private void backtrack(int[] nums, int start, List<Integer> current, List<List<Integer>> result) {
        result.add(new ArrayList<Integer>(current));

        for (int i = start; i < nums.length; i++) {
            current.add(nums[i]);
            backtrack(nums, i + 1, current, result);
            current.remove(current.size() - 1);
        }
    }

    // SOLUTION
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        backtrack(nums, 0, current, result);
        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] nums = {1,2,3};

        // OUTPUT
        var result = o.subsets(nums);

        System.out.print("["); for (var x : result) {
            System.out.print("["); for (var y : x) {
                System.out.print(y + ",");
            } System.out.print("\b],");
        } System.out.println("\b]");
    }
}
